feat: Matching OCPP 2.0.1 payload types to OCPP 2.0.1 json schema

ajmirsky opened the pull request:
This brings @malsyned #562 up-to-date with base branch and fixes linting errors.

*Justification*

Payload types now match the json schema for OCPP2.0.1 that are also included in this library. allows for static type checking of payloads.

*Breaking changes*

To verify that there are no breaking changes, the unit test included in #684 is used to prevent regression with the nested model datatypes.

```
from ocpp.v201.call import BootNotification

def process_boot_notification(payload: BootNotification):
    vendor_name = payload.charging_station['vendor_name']


payload =  json.loads('{"charging_station": {"vendor_name": "OC", "model": "OC-X", "modem": {"iccid": "1234567890", "imsi": "1234567890"}, "serial_number": "123456", "firmware_version": "1.0.0"}, "reason": "PowerUp", "custom_data": null}')

boot_notification_payload = BootNotification(**payload)
type(boot_notification_payload.charging_station)  # <class 'dict'>

process_boot_notification(boot_notification_payload)
```

To hydrate with nested datatypes, use [json-dataclasses](https://github.com/lidatong/dataclasses-json/blob/master/dataclasses_json/core.py#L156)

ajmirsky commented:
Included same unit test for v2.0.1 datatypes and enums to make sure that nested model type changes doesn't alter functionality.

jainmohit2001 commented:
```python
@on(Action.TransactionEvent)
def on_transaction_event(self, transaction_info: dict, *args, **kwargs):
    transaction_id = transaction_info.get("transaction_id", None)
    # I can access all the data inside transaction_info using the `.get` method.
    meter_value = kwargs.get("meter_value", [])
    # I can access meter values using the same .get method.
```



My question is, whether the following will work after the changes in dataclasses?

```python
@on(Action.TransactionEvent)
def on_transaction_event(self, transaction_info: TransactionEvent, *args, **kwargs):
    transaction_id = transaction_info.get("transaction_id", None)
    meter_value = kwargs.get("meter_value", [])
```

jainmohit2001 commented:
NVM. The question is resolved.
Upon carefully inspecting the `_handle_call` function in `charge_point.py`, it's clear that the data passed to `@on` handlers is simply dictionary. We don't convert the incoming data to data classes here.

ajmirsky commented:
@jainmohit2001 yup. the type is just a hint in python; it won't cast the object to a new type. so as long as you interact with the `transaction_info` as if it were a dict, then you're ok. 

The change you're describing can be made now, without the dataclass changes in this pull request, since [TransactionEvent](https://github.com/mobilityhouse/ocpp/blob/master/ocpp/v201/call.py#L449) already exists.

```
@on(Action.TransactionEvent)
def on_transaction_event(self, transaction_info: TransactionEvent, *args, **kwargs):
    transaction_id = transaction_info.get("transaction_id", None)
    meter_value = kwargs.get("meter_value", [])
```

Before or after this pull request, a static type checker or IDE should show a warning. With the type hint of `TransactionEvent`, the checkers will expect you to access `transaction_info` not with `get` but by accessing the attribute directly: `transaction_info.transaction_id`. But, as you point out, `charge_point.py` still sends as a dict.

ajmirsky commented:
@jainmohit2001 now that your question has been answered, will this PR be reviewed/approved for inclusion in the next release?

jainmohit2001 commented on ocpp/v201/enums.py:
```diff
@@ -721,7 +741,7 @@ class HashAlgorithmType(StrEnum):
 
 
-class IdTokenType(StrEnum):
+class IdTokenEnumType(StrEnum):
```
Can you help me understand why are we renaming `IdTokenType` and `UnitOfMeasureType` to `IdTokenEnumType` and `StandardizedUnitsOfMeasureType`?

jainmohit2001 replied:
This is related to OCPP nomenclature

jainmohit2001 commented:
As per the official OCPP 2.0.1 Part 2: Specification document, all the enum types have `Enum` in their names.
Screenshot for your reference:
<img width="896" alt="image" src="https://github.com/user-attachments/assets/60e25eaf-a3f8-4b63-b67b-a7587155e0c9">

The current enum declarations in OCPP package don't align with this nomenclature.

jainmohit2001 commented:
There is a mismatch between the official `ConnectorEnumType` and the `ConnectorType` defined. The python version has additional types as compared to the official `ConnectorEnumType`
```json
"ConnectorEnumType": {
	  "description": "This field specifies the connector type.\r\n",
	  "javaType": "ConnectorEnum",
	  "type": "string",
	  "additionalProperties": false,
	  "enum": [
		"cCCS1",
		"cCCS2",
		"cG105",
		"cTesla",
		"cType1",
		"cType2",
		"s309-1P-16A",
		"s309-1P-32A",
		"s309-3P-16A",
		"s309-3P-32A",
		"sBS1361",
		"sCEE-7-7",
		"sType2",
		"sType3",
		"Other1PhMax16A",
		"Other1PhOver16A",
		"Other3Ph",
		"Pan",
		"wInductive",
		"wResonant",
		"Undetermined",
		"Unknown"
	  ]
	},
```

```python

class ConnectorType(StrEnum):
    """
    Allowed values of ConnectorCode.
    """

    # Combined Charging System 1 (captive cabled) a.k.a. Combo 1
    c_ccs1 = "cCCS1"
    # Combined Charging System 2 (captive cabled) a.k.a. Combo 2
    c_ccs2 = "cCCS2"
    # ChaoJi: New CHAdeMO connector harmonised with GB/T
    c_chao_ji = "cChaoJi"
    # JARI G105-1993 (captive cabled) a.k.a. CHAdeMO
    c_g105 = "cG105"
    # GB/T: Chinese DC charging connector
    c_gbt = "cGBT"
    # Tesla Connector (captive cabled)
    c_tesla = "cTesla"
    # IEC62196-2 Type 1 connector (captive cabled) a.k.a. J1772
    c_type1 = "cType1"
    # IEC62196-2 Type 2 connector (captive cabled) a.k.a. Mennekes connector
    c_type2 = "cType2"
    # 16A 1 phase IEC60309 socket
    s309_1p_16a = "s309-1P-16A"
    # 32A 1 phase IEC60309 socket
    s309_1p_32a = "s309-1P-32A"
    # 16A 3 phase IEC60309 socket
    s309_3p_16a = "s309-3P-16A"
    # 32A 3 phase IEC60309 socket
    s309_3p_32a = "s309-3P-32A"
    # UK domestic socket a.k.a. 13Amp
    s_bs1361 = "sBS1361"
    # CEE 7/7 16A socket. May represent 7/4 & 7/5 a.k.a Schuko
    s_cee_7_7 = "sCEE-7-7"
    # IEC62196-2 Type 2 socket a.k.a. Mennekes socket
    s_type2 = "sType2"
    # IEC62196-2 Type 3 socket a.k.a. Scame
    s_type3 = "sType3"
    # Reverse pantograph
    opp_charge = "OppCharge"
    # Other single phase (domestic) sockets not mentioned above, rated at
    # no more than 16A. CEE7/17, AS3112, NEMA 5-15, NEMA 5-20, JISC8303,
    # TIS166, SI 32, CPCS-CCC, SEV1011, etc.
    other_1ph_max_16a = "Other1PhMax16A"
    # Other single phase sockets not mentioned above (over 16A)
    other_1ph_over_16a = "Other1PhOver16A"
    # Other 3 phase sockets not mentioned above. NEMA14-30, NEMA14-50
    other_3ph = "Other3Ph"
    # Pantograph connector
    pan = "Pan"
    # Wireless inductively coupled connection (generic)
    w_inductive = "wInductive"
    # Wireless resonant coupled connection (generic)
    w_resonant = "wResonant"
    # Yet to be determined (e.g. before plugged in)
    undetermined = "Undetermined"
    # Unknown & not determinable
    unknown = "Unknown"
```

jainmohit2001 commented:
Mismatch between DataEnumType and DataType. Python version has an extra entry: `password_string`
```json
"DataEnumType": {
	  "description": "Data type of this variable.\r\n",
	  "javaType": "DataEnum",
	  "type": "string",
	  "additionalProperties": false,
	  "enum": [
		"string",
		"decimal",
		"integer",
		"dateTime",
		"boolean",
		"OptionList",
		"SequenceList",
		"MemberList"
	  ]
	},
```

```python
class DataType(StrEnum):
    """
    DataEnumType is used by CommonVariableCharacteristicsType
    """

    string = "string"
    decimal = "decimal"
    integer = "integer"
    date_time = "dateTime"
    boolean = "boolean"
    option_list = "OptionList"
    sequence_list = "SequenceList"
    member_list = "MemberList"
    password_string = "passwordString"
```

jainmohit2001 commented:
Mismatch between the official `MessageStateEnumType` and the defined `MessageStateType`. The python version is missing the entry: `Unavailable`

<details>
<summary>MessageStateEnumType</summary>

```json
"MessageStateEnumType": {
	  "description": "If provided the Charging Station shall return Display Messages with the given state only. \r\n",
	  "javaType": "MessageStateEnum",
	  "type": "string",
	  "additionalProperties": false,
	  "enum": [
		"Charging",
		"Faulted",
		"Idle",
		"Unavailable"
	  ]
	}
```
</details>

<details>
<summary>MessageStateType</summary>

```python
class MessageStateType(StrEnum):
    """
    State of the Charging Station during which a message SHALL be displayed.
    """

    charging = "Charging"
    faulted = "Faulted"
    idle = "Idle"
```

</details>

jainmohit2001 commented:
### Standardized units of measure
To maintain consistency with the OCPP nomenclature, we need to make the following changes:

- [ ] Rename `UnitOfMeasureType` to `StandardizedUnitsofMeasureEnumType`
- [ ] Re order the declaration of fields and match them with OCPP docs
- [ ] Fix values for entries:
    - [ ] `b`: `Bytes` -> `B`

jainmohit2001 commented:
### StatusInfoReasonType

- [ ] Fix the values for the following entries:
    - [ ] `invalid_message_sequence`: `InvalidMessageSequence` -> `InvalidMessageSeq`

jainmohit2001 commented:
### GenericVariableName

- [ ] We should rename `GenericVariableName` to `StandardizedVariableName`
- [ ] Add entry for `ISO15118EvseId` right after `Interval`

jainmohit2001 commented:
### AuthCacheCtrlrVariableName

- Additional entry for `Storage`

jainmohit2001 commented:
### DeviceDataCtrlrVariableName

- Additional entries present for `ConfigurationValueSize` and `ReportingValueSize`, which are not part of official list of variable names.

jainmohit2001 commented:
### DisplayMessageCtrlrVariableName

- Additional entry present for `SupportedPriorities`, which is not part of official list of variable names.

jainmohit2001 commented:
### OCPPCommCtrlrVariableName

- Additional entry present for `ResetRetries`.

jainmohit2001 commented on ocpp/v201/datatypes.py:
```diff
@@ -555,7 +555,7 @@ class UnitOfMeasureType:
     """
 
-    unit: Optional[str] = None
+    unit: Optional[enums.StandardizedUnitsOfMeasureType] = None
```
<img width="830" alt="image" src="https://github.com/user-attachments/assets/c72e8e7c-f4fe-42c0-9e7f-740609d4b342">

Will this change adhere to the specs from OCPP docs? How can I define `custom` unit

ajmirsky replied:
I find this part of the spec confusing. They emphasize *SHALL* use from the list, which implies those are the only valid ones. Then go on to say that one can create a custom unit?

Either way, I've updated the type hint to:
`unit: Optional[Union[enums.StandardizedUnitsOfMeasureEnumType, str]] = None

jainmohit2001 commented:
@ajmirsky, correct me if I am wrong, but the changes in this PR for the data classes corresponding to OCPP call and call result only supports nesting upto level 1. For level two, you still need to access the attribute using `.get()` method or just using the key: `parent_type.child_type['key']`

jainmohit2001 commented:
Hi @ajmirsky, please have a look at my comments whenever you get the time.

ajmirsky commented:
> As per the official OCPP 2.0.1 Part 2: Specification document, all the enum types have `Enum` in their names. Screenshot for your reference: <img alt="image" width="896" src="https://private-user-images.githubusercontent.com/45727354/394128266-60e25eaf-a3f8-4b63-b67b-a7587155e0c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwMTg0NDgsIm5iZiI6MTczNDAxODE0OCwicGF0aCI6Ii80NTcyNzM1NC8zOTQxMjgyNjYtNjBlMjVlYWYtYTNmOC00YjYzLWI2N2ItYTc1ODcxNTVlMGM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEyVDE1NDIyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhNmE2NWVmYWNiNTRkNmFjOTEyYjQ5ZjMyZTFkODUxMjRiODZkNzk2MTQ3ZjM3MDNhNTVmNmE5NTU4OTU1ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5OH0inwk3RL5VYwG2v5aEiwHN5_Hr5aTFs3GMbGJnrg">
> 
> The current enum declarations in OCPP package don't align with this nomenclature. In my opinion, we should update the enum names to align with the official nomenclature and keep the old ones with deprecation warnings. In the next major release we can remove the deprecated warnings as well. Let me know your thoughts here. @ajmirsky @proelke

fixed.

ajmirsky commented:
> @ajmirsky, correct me if I am wrong, but the changes in this PR for the data classes corresponding to OCPP call and call result only supports nesting upto level 1. For level two, you still need to access the attribute using `.get()` method or just using the key: `parent_type.child_type['key']`. Can you help me understand how should I benefit from strong typings in this case? Let's take the same example that you have started with:
> 
> ```python
> from ocpp.v201.call import BootNotification
> 
> def process_boot_notification(payload: BootNotification):
>     vendor_name = payload.charging_station['vendor_name']  # This is not strongly typed?
> 
>     # vendor_name = payload.charging_station.vendor_name  # I cannot do this without using another library as you mentioned
> 
> 
> payload =  json.loads('{"charging_station": {"vendor_name": "OC", "model": "OC-X", "modem": {"iccid": "1234567890", "imsi": "1234567890"}, "serial_number": "123456", "firmware_version": "1.0.0"}, "reason": "PowerUp", "custom_data": null}')
> 
> boot_notification_payload = BootNotification(**payload)
> type(boot_notification_payload.charging_station)  # <class 'dict'>
> 
> process_boot_notification(boot_notification_payload)
> ```

This is the trade off of Python's implementation of types; they are type _hints_ for static type checking and not to be confused with a strongly typed language with Java and C. At runtime, python is still loosely typed at runtime the `BootNotification` payload can be created "loosely" without following the typed hints.

From a python perspective, the type hinting makes this correct:

```
boot_notification = BootNotification(
    charging_station=ChargingStationType(
        model="Station Model X",
        vendor_name="Vendor ABC",
        serial_number="SN123456",
        modem=ModemType(iccid="89001234567890123456", imsi="123456789012345"),
        firmware_version="1.2.3",
    ),
    reason=BootReasonEnumType.power_up,
)
```

In the end, it's really a mismatch about how json is translated into python. Similar to how we need the functions `camel_case_to_snake_case` (and its inverse) or `remove_nones` to make the datatypes "pythonic" while also complying with the json schema, one needs a similar translation function to "properly" create the payloads which follow the type hints. This happens to already be available in a third party library. 

It could be argued that the ocpp library should supply this utility function without relying on a third party library. If so, I'd be happy to create our own function to decode json into the proper nested dataclasses. In theory, it should be part of the `unpack` function, but that would be a breaking change, requiring everyone to access their payloads in a new manner. So including it as an optional helper function makes the most sense (I'll do this on a separate branch and separate PR since this PR isn't dependent on it).

Should also highlight that `charge_point.py` is just one implementation. And, for better or worse, the implementation decided not to follow type hints when creating payloads. I think we should be creating an ocpp library that is true to the OCPP specification; and not one that is tied to particular design decisions associated with one particular implementation of a charge point.

In the end, I think we're in a good position because we can satisfy both; we can follow the spec for those who want to implement using type hints and also make the changes without breaking code already written that is loosely typed.

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -533,7 +553,7 @@ class DataTransferStatusType(StrEnum):
 
 
-class DataType(StrEnum):
+class DataEnumEnumType(StrEnum):
```
oops : )

fixed

ajmirsky commented:
> ### Standardized units of measure
> To maintain consistency with the OCPP nomenclature, we need to make the following changes:
> 
> * [ ]  Rename `UnitOfMeasureType` to `StandardizedUnitsofMeasureEnumType`

There is a `UnitOfMeasureType` which is a datatype (2.49). This datatype has a field `unit` which is of type `StandardizedUnitsofMeasureEnumType` (based on Appendix 2 of OCPP 2.0.1 Part 2). 

> * [ ]  Re order the declaration of fields and match them with OCPP docs

everything, i believe, matches the json schema; which comes directly from OCPP. so I think we're ok if we just follow the ordering of the schema

> * [ ]  Fix values for entries:
>   
>   * [ ]  `b`: `Bytes` -> `B`
>

changed

> ### StatusInfoReasonType
> * [ ]  Fix the values for the following entries:
>   
>   * [ ]  `invalid_message_sequence`: `InvalidMessageSequence` -> `InvalidMessageSeq`
> 

changed

> ### GenericVariableName
> * [ ]  We should rename `GenericVariableName` to `StandardizedVariableName`

changed

> * [ ]  Add entry for `ISO15118EvseId` right after `Interval`

added

> ### TxCtrlrVariableName
> * [ ]  Additional entry present for `ChargingTime`

removed

> ### ChargingStationVariableName
> * [ ]  Additional entry present for `Identity`

removed

> ### ConnectedEVVariableName
> * [ ]  A lot of missing values from `3.2.13. ConnectedEV` and additional entries which are not part of official OCPP list of variable names.

updated


> ### LocalEnergyStorageVariableName
> * [ ]  Additional entry present for `Capacity`

removed

ajmirsky commented:
@jainmohit2001 thanks for the careful review and double check of things. please let me know if I've missed anything else.

jainmohit2001 commented:
Thanks for updating the PR quickly. I'll be reviewing the whole PR once again soon.

jainmohit2001 reviewed (changes requested):
Added my review for enums and datatypes.
Will take up call and call_result next.

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -10,6 +10,26 @@ class StrEnum(str, Enum):  # pragma: no cover
         pass  # pragma: no cover
 
 
+class DeprecatedEnumWrapper:
```
Sounds good.

jainmohit2001 replied:
@proelke @drc38 @jerome-benoit, let me know if this makes sense.

jerome-benoit replied:
I agree.

jainmohit2001 replied:
Resolving this conversation. Will pick up migration guide soon!

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -354,7 +374,7 @@ class ChargingProfileStatus(StrEnum):
     rejected = "Rejected"
```
fixed

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -736,7 +756,10 @@ class IdTokenType(StrEnum):
 
 
-class InstallCertificateStatusType(StrEnum):
+IdTokenType = DeprecatedEnumWrapper(IdTokenEnumType, "IdTokenType")
```
done

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -1321,7 +1344,7 @@ class VPNType(StrEnum):
 # DataTypes
```
fixed

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -1364,7 +1387,12 @@ class UnitOfMeasureType(StrEnum):
 
 
-class StatusInfoReasonType(StrEnum):
+UnitOfMeasureType = DeprecatedEnumWrapper(
```
done

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -1321,7 +1344,7 @@ class VPNType(StrEnum):
 # DataTypes
```
removed.

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -1592,6 +1620,7 @@ class GenericVariableName(StrEnum):
     impedance = "Impedance"
     imsi = "IMSI"
     interval = "Interval"
+    iso_15118_evse_id = "ISO15118EvseId"
```
done

ajmirsky commented on ocpp/v201/enums.py:
```diff
@@ -2152,26 +2179,50 @@ class ConnectedEVVariableName(StrEnum):
     charging_complete_full = "ChargingCompleteFull"
 
+    # Status values
+    battery_over_voltage = " BatteryOvervoltage"
```
battery_overvoltage and battery_undervoltage fixed. I think `charging_current_deviation = " ChargingCurrentDeviation"` is correct.

jainmohit2001 replied:
<img width="841" alt="image" src="https://github.com/user-attachments/assets/a101db52-1c38-4b8f-b927-8a85025569c8" />

Can you help me understand where the extra space is coming from?

ajmirsky replied:
i extracted the values from the pdf and didn't use `rstrip()` : ( 

i'll fix

jainmohit2001 commented on ocpp/v201/call_result.py:
```diff
@@ -323,62 +327,62 @@ class SecurityEventNotification:
 class SetMonitoringBase:
-    status: str
-    status_info: Optional[Dict] = None
+    status: enums.GenericStatusEnumType
```
<img width="841" alt="image" src="https://github.com/user-attachments/assets/dfe7176b-97ab-482d-93d7-7213e6f508fc" />

Should we use `GenericDeviceModelStatusEnumType` or `GenericStatusEnumType`?

ajmirsky replied:
changed to `GenericDeviceModelStatusEnumType`

jainmohit2001 reviewed (changes requested):
Completed review for call and call_result

jainmohit2001 commented:
@ajmirsky You have included the tests for data types in this PR and in https://github.com/mobilityhouse/ocpp/pull/684 as well. I think we can close https://github.com/mobilityhouse/ocpp/pull/684 and keep the tests in this PR itself.

jerome-benoit commented on ocpp/v201/enums.py:
```diff
@@ -2152,26 +2151,50 @@ class ConnectedEVVariableName(StrEnum):
     charging_complete_full = "ChargingCompleteFull"
 
+    # Status values
+    battery_overvoltage = " BatteryOvervoltage"
```
```suggestion
    battery_overvoltage = "BatteryOvervoltage"
```

ajmirsky replied:
fixed

jerome-benoit commented on ocpp/v201/enums.py:
```diff
@@ -2152,26 +2151,50 @@ class ConnectedEVVariableName(StrEnum):
 
+    # Status values
+    battery_overvoltage = " BatteryOvervoltage"
+    battery_undervoltage = " BatteryUndervoltage"
```
```suggestion
    battery_undervoltage = "BatteryUndervoltage"
```

ajmirsky replied:
fixed

jerome-benoit commented on ocpp/v201/enums.py:
```diff
@@ -2152,26 +2151,50 @@ class ConnectedEVVariableName(StrEnum):
+    # Status values
+    battery_overvoltage = " BatteryOvervoltage"
+    battery_undervoltage = " BatteryUndervoltage"
+    charging_current_deviation = " ChargingCurrentDeviation"
```
```suggestion
    charging_current_deviation = "ChargingCurrentDeviation"
```

ajmirsky replied:
fixed

ajmirsky commented:
> Hi @ajmirsky, please look at my comments whenever you get the time. I think we are good to go for this PR apart from the comments I have mentioned.

Apologies for the delay. They should now be resolved.

ajmirsky commented:
> @ajmirsky You have included the tests for data types in this PR and in #684 as well. I think we can close #684 and keep the tests in this PR itself.

closed #684

ajmirsky commented:
@jainmohit2001 since we're removing the deprecated enums, should we also be removing the deprecated call and call_result Payloads? Or is that scheduled for a future release?

jainmohit2001 commented:
> @jainmohit2001 since we're removing the deprecated enums, should we also be removing the deprecated call and call_result Payloads? Or is that scheduled for a future release?

A part of that is being done in #694

jainmohit2001 commented:
@proelke, have a look at the comments in this PR and share your thoughts. I have gone through the changes thoroughly and we are good for merge.

proelke commented:
My only feedback would be the use of importing the module and not each class. This diverges from how it is done for the 1.6 code. I'm personally more of a fan how it's done in the 1.6 code base vs what this PR introduces but they should be consistent.

jainmohit2001 commented:
> My only feedback would be the use of importing the module and not each class. This diverges from how it is done for the 1.6 code. I'm personally more of a fan how it's done in the 1.6 code base vs what this PR introduces but they should be consistent.

Considering the fact that OCPP 2.0.1 have a large number of datatypes and enums, importing each class instead of the whole module would only bring long import lines. I agree that we should be consistent in imports. We can create a separate PR for that and update the OCPP 1.6 imports if that works.
